File: Leetcode/UniquePathsII.cpp
class Solution {
public:
    int uniquePathsWithObstacles(vector<vector<int>>& g) {
        int n = g.size(); 
        if(n == 0) return 0;
        int m = g[0].size();
        int dp[n + 1][m + 1];
        memset(dp, 0, sizeof dp);
        dp[1][0] = 1;
        for(int i = 1; i <= n; ++i){
            for(int j = 1; j <= m; ++j){
                if(g[i - 1][j - 1] == 1) dp[i][j] = 0;
                else dp[i][j] = dp[i - 1][j] + dp[i][j - 1];
            }
        }
        return dp[n][m];
    }
};

File: Leetcode/UniquePathsII.rs
// https://leetcode.com/problems/unique-paths-ii

struct Solution {}

impl Solution {
    pub fn unique_paths_with_obstacles(obstacle_grid: Vec<Vec<i32>>) -> i32 {
        let n = obstacle_grid.len();
        let m = obstacle_grid[0].len();
        let mut dp = vec![vec![0; m]; n];
        dp[0][0] = (obstacle_grid[0][0] != 1) as i32;

        for i in 0..n {
            for j in 0..m {
                if i + 1 < n && obstacle_grid[i + 1][j] != 1 {
                    dp[i + 1][j] += dp[i][j];
                }
                if j + 1 < m && obstacle_grid[i][j + 1] != 1 {
                    dp[i][j + 1] += dp[i][j];
                }
            }
        }

        dp[n - 1][m - 1]
    }
}
